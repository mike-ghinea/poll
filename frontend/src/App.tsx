import Poll from "./components/Poll/Poll";

const App = () => {
  return <Poll />;
};
export default App;
