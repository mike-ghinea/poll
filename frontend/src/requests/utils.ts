export const apiRoot = "http://localhost:8080";
export const pollDomain = "/api/poll";

export interface PollOption {
  id: string;
  text: string;
}

export interface Poll {
  id: string;
  question: string;
  options: PollOption[];
}

export interface PollOptionWithVoteCount extends PollOption {
  count: number;
}
