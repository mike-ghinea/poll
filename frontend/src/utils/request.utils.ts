export const apiRoot = "http://localhost:8080";
export const pollDomain = "/api/poll";
