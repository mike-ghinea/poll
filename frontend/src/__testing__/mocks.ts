import { Poll } from "../requests/models";

export const mockPoll: Poll = {
  id: "id",
  question: "Why?",
  options: [],
};
