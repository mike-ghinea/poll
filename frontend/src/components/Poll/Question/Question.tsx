import { Poll } from "../../../requests/utils";
import { postVote } from "../../../requests/vote";
import { ErrorMessage, SIZE } from "../../ErrorMessage/ErrorMessage";
import { LoadingBox } from "../../LoadingBox/LoadingBox";
import s from "../Poll.styles";
import React, { useState } from "react";

const Question: React.FC<{
  poll: Poll;
  showResults: () => void;
}> = ({ poll, showResults }) => {
  const [selectedOption, setSelectedOption] = React.useState<string>();

  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<string>();

  const onSubmit = () => {
    if (!selectedOption) return;
    setIsLoading(true);
    setError(undefined);
    postVote(poll.id, selectedOption)
      .then(() => showResults())
      .catch(() =>
        setError(
          "Something went wrong while submitting your vote. Please try again later.",
        ),
      )
      .finally(() => setIsLoading(false));
  };

  if (isLoading) return <LoadingBox />;

  if (error) return <ErrorMessage size={SIZE.TITLE} message={error} />;

  return (
    <>
      <s.Header>{poll.question}</s.Header>
      <s.OptionWrapper>
        {poll.options.map((option) => (
          <s.PollOption
            key={option.id}
            onClick={(e) => {
              e.preventDefault();
              setSelectedOption(option.id);
            }}
            $hightlight={option.id === selectedOption}
          >
            {option.text}
          </s.PollOption>
        ))}
      </s.OptionWrapper>
      <s.SubmitButton
        role="submit"
        onClick={onSubmit}
        disabled={!selectedOption}
        aria-disabled={!selectedOption}
      >
        Submit
      </s.SubmitButton>
    </>
  );
};

export default Question;
