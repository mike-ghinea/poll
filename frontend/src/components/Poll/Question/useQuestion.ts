import { useState } from "react";
import { postVote } from "../../../requests/vote";

export const useQuestion = (pollId: string, showResults: () => void) => {
  const [selectedOption, setSelectedOption] = useState<string>();
  const updateSelectedOption = (option: string) => setSelectedOption(option);

  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<string>();

  const onSubmit = () => {
    if (!selectedOption) return;
    setIsLoading(true);
    setError(undefined);
    postVote(pollId, selectedOption)
      .then(() => showResults())
      .catch(() =>
        setError(
          "Something went wrong while submitting your vote. Please try again later.",
        ),
      )
      .finally(() => setIsLoading(false));
  };

  return {
    isLoading,
    error,
    selectedOption,
    onSubmit,
    updateSelectedOption,
  };
};
