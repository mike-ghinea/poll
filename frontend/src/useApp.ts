import { useEffect, useState } from "react";
import { getActivePoll } from "./requests/poll";
import { Poll } from "./requests/models";

export const useApp = () => {
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<string>();
  const [activePoll, setActivePoll] = useState<Poll>();

  useEffect(() => {
    setIsLoading(true);
    setError(undefined);
    getActivePoll()
      .then((res) => setActivePoll(res.data))
      .catch((e: { status: number }) => {
        if (e.status === 404)
          setError(
            "There are no active polls at the moment. Please check in later.",
          );
        else
          setError(
            "We are not able to get the active poll at the moment. Please try again later.",
          );
      })
      .finally(() => setIsLoading(false));
  }, []);

  return {
    isLoading,
    error,
    activePoll,
  };
};
