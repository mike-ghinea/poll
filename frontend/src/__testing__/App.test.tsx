import App from "../App";
import { render } from "../utils/test.utils";

import * as appHook from "../useApp";
import * as questionHook from "../components/Poll/Question/useQuestion";
import { mockPoll } from "./mocks";

const useAppSpy = jest.spyOn(appHook, "useApp");
const useQuestionSpy = jest.spyOn(questionHook, "useQuestion");

describe("App", () => {
  beforeEach(() => {
    useAppSpy.mockReturnValue({
      isLoading: false,
      error: undefined,
      activePoll: mockPoll,
    });
    useQuestionSpy.mockReturnValue({
      isLoading: false,
      error: undefined,
      selectedOption: undefined,
      onSubmit: jest.fn(),
      updateSelectedOption: jest.fn(),
    });
  });
  afterEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
  });
  it("should render correctly", () => {
    const { baseElement } = render(<App />);
    expect(baseElement).toMatchSnapshot();
  });
});
