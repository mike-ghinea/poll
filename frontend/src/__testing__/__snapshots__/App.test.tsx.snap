// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<body>
  <div>
    <div
      class="sc-FEMpB jnEXiv"
    >
      <h1
        class="sc-dycHyt epsnkE"
      >
        Why?
      </h1>
      <div
        class="sc-gjcoXW eBlsSB"
      />
      <button
        aria-disabled="true"
        class="sc-jbAkgO dsqhJQ"
        disabled=""
        role="submit"
      >
        Submit
      </button>
    </div>
  </div>
</body>
`;
